src/System.Private.ServiceModel/src/System/ServiceModel/Channels/Connection.cs

```diff
@@ -31,6 +31,62 @@ public interface IConnection
         int EndRead();
     }
 
+    public static class IConnectionExtensions
+    {
+        public static async Task WriteAsync(this IConnection connection, byte[] buffer, int offset, int size, bool immediate, TimeSpan timeout)
+        {
+            var tcs = new TaskCompletionSource<bool>();
+            var result = connection.BeginWrite(buffer, offset, size, immediate, timeout, OnIoComplete, tcs);
+            if (result == AsyncCompletionResult.Completed)
+            {
+                tcs.SetResult(true);
+            }
+
+            await tcs.Task;
+            connection.EndWrite();
+        }
+
+        public static async Task<int> ReadAsync(this IConnection connection, int offset, int size, TimeSpan timeout)
+        {
+            // read ACK
+            var tcs = new TaskCompletionSource<bool>();
+            //ackBuffer
+
+            var result = connection.BeginRead(offset, size, timeout, OnIoComplete, tcs);
+            if (result == AsyncCompletionResult.Completed)
+            {
+                tcs.SetResult(true);
+            }
+
+            await tcs.Task;
+            int ackBytesRead = connection.EndRead();
+            return ackBytesRead;
+        }
+
+        public static async Task<int> ReadAsync(this IConnection connection, byte[] buffer, int offset, int size, TimeSpan timeout)
+        {
+            int ackBytesRead = await connection.ReadAsync(offset, size, timeout);
```

Comment:
Need a validation check to make sure buffer!=null and buffer.Length >= offset + size. See some of our stream implementations such as ConnectionStream for an example.